### PR TITLE
Country level conversion

### LINKED ISSUE #559: Scrape and cross-check JHU county-level data
Just got added: https://github.com/CSSEGISandData/COVID-19/blob/master/csse_covid_19_data/csse_covid_19_time_series/time_series_covid19_confirmed_US.csv

### LINKED ISSUE #659: Duplicated US Data
It appears data for April 3 is duplicated into April 4 for many US locations.

### LINKED ISSUE #675: USA country name changed to iso1:US, similar for states
In timeseries data and snapshot, the country name for USA is has changed to iso1:US and states also have a changed from state abbreviation to iso1:US-WA vs WA, for example. Could this be changed back to USA and just state abbreviation, for consistency?

image

### PR DESCRIPTION
**Pull request by hyperknot**
## Summary

Scrape and cross-check JHU county level data (in addition to all other data), closes #559
Fixes duplicated data due to `United States` vs `iso1:US`, closes #659 
Fix `country` and `countryId`, closes #675

Migrating all scrapers to use country level IDs. 

Scrapers didn't use ids.

Scrapers use ids for country, almost all scrapers use id for state and county.

## Changes

- JHU / JHU-USA major rewrite.
- JHU got down froom 330 to 170 lines of code. I removed all kind of magic functions, point-in-polygon detection. It's now only using the official ISO / FIPS mapping list by JHU.
- Renamed all 3 letter iso countries to 2 letter ones, with iso1: prefix
- Manually fixed to use state ID codes:
 - Germany
 - Spain
 - Canada

Also, because of the search & replace, all scraper files have been updated in one line.

## Additional notes

- We should add ids to timeseries.csv
- ids could be visible more "left" in columns, next to the names I believe. I think many users don't even realize we have them now.

### CODE REVIEW

**Comment by hyperknot** [comment]
@lazd

**Comment by hyperknot** [comment]
I forgot to mention we did many parts to gether with @lazd.

File: package.json [hunk 1]
```diff
@@ -41,7 +41,7 @@
     "ajv": "^6.12.0",
     "cheerio": "^1.0.0-rc.3",
     "cheerio-tableparser": "^1.0.1",
-    "country-levels": "https://github.com/hyperknot/country-levels/releases/download/v1.2.0/export_q7.tgz",
+    "country-levels": "https://github.com/hyperknot/country-levels-releases/releases/download/v.1.4.0/export_q7.tgz",
```
Comment (camjc) [inline package.json:44]: ? is this legit?

Comment (lazd) [reply package.json:44]: So legit

File: src/shared/scrapers/CA/index.js [hunk 2]
```diff
@@ -1,15 +1,33 @@
...
 
+const countryLevelMap = {
+  Canada: 'iso1:CA',
+  'British Columbia': 'iso2:CA-BC',
+  'New Brunswick': 'iso2:CA-NB',
+  'Newfoundland and Labrador': 'iso2:CA-NL',
+  'Northwest Territories': 'iso2:CA-NT',
+  'Nova Scotia': 'iso2:CA-NS',
+  'Prince Edward Island': 'iso2:CA-PE',
+  'Repatriated travellers': '-',
```
Comment (camjc) [inline src/shared/scrapers/CA/index.js:19]: We'll handle these correctly (not even sure what correctly would do here, still add to the canada total?)?

File: src/shared/scrapers/CA/index.js [hunk 3]
```diff
@@ -94,7 +125,8 @@ const scraper = {
         throw new Error(`Timeseries does not contain a sample for SCRAPE_DATE ${scrapeDateString}`);
       }
 
-      regions.push(transform.sumData(regions));
+      // no need to do this as they provide as the summary data
```
Comment (camjc) [inline src/shared/scrapers/CA/index.js:128]: I've been asserting the summed === the total scraped from the site, may be useful, not sure.

File: src/shared/scrapers/DE/DE-BB/index.js [hunk 4]
```diff
@@ -3,13 +3,13 @@ import defaultScraperDEU, { sharedSchema } from '../_shared.js';
...
-  state: 'DE-BB', // ISO 3166 notation
+  country: 'iso1:DE',
+  state: 'iso2:DE-BB', // ISO 3166 notation
   scraper: defaultScraperDEU,
   _rowToResult: row => {
     return {
-      cases: parse.number(row[`${scraper.state}_cases`]),
-      deaths: parse.number(row[`${scraper.state}_deaths`]),
+      cases: parse.number(row[`${scraper.state.slice(5)}_cases`]),
+      deaths: parse.number(row[`${scraper.state.slice(5)}_deaths`]),
```
Comment (camjc) [inline src/shared/scrapers/DE/DE-BB/index.js:12]: Why?

File: src/shared/scrapers/ES/index.js [hunk 5]
```diff
@@ -1,12 +1,35 @@
...
 import datetime from '../../lib/datetime/iso/index.js';
 import * as fetch from '../../lib/fetch/index.js';
 import * as transform from '../../lib/transform.js';
 import * as parse from '../../lib/parse.js';
 
 const { looksLike } = datetime;
 
+const countryLevelMap = {
+  Andalucía: 'iso2:ES-AN',
+  Aragón: 'iso2:ES-AR',
```
Comment (camjc) [inline src/shared/scrapers/ES/index.js:11]: is it worth deburring these before matching?

File: src/shared/scrapers/JHU-USA.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+import * as parse from '../lib/parse.js';
+import * as geography from '../lib/geography/index.js';
+import * as transform from '../lib/transform.js';
+import datetime from '../lib/datetime/index.js';
+import maintainers from '../lib/maintainers.js';
+
+// Set county to this if you only have state data, but this isn't the entire state
+const UNASSIGNED = '(unassigned)';
+
+function findCountryLevelID(fips) {
```
Comment (camjc) [inline src/shared/scrapers/JHU-USA.js:12]: is this too simple to be useful as an abstraction layer? why not just call fipsCodes[fips]?

File: src/shared/scrapers/JHU-USA.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+      // Sum the whole country
+      regions.push(transform.sumData(regions));
+
+      // Sum individual states
+      for (const [state, locations] of Object.entries(stateLocations)) {
+        regions.push(transform.sumData(locations, { state }));
+      }
+
+      // remove unassigned counties one we summed them up
```
Comment (camjc) [inline src/shared/scrapers/JHU-USA.js:124]:
```suggestion
      // remove unassigned counties once we have summed them up
```

File: src/shared/scrapers/US/KY/index.js [hunk 8]
```diff
@@ -21,7 +21,7 @@ const scraper = {
       name: 'Jordan Holt',
       email: 'jordholt@gmail.com',
       github: 'Jord-Holt',
-      country: 'USA',
+      country: 'iso1:US',
```
Comment (camjc) [inline src/shared/scrapers/US/KY/index.js:24]: Is this legit?

File: src/shared/scrapers/US/NYT.js [hunk 9]
```diff
@@ -25,26 +25,25 @@ const scraper = {
     const data = await fetch.csv(this.url, false);
 
     // FIXME when we roll out new TZ support!
```
Comment (camjc) [inline src/shared/scrapers/US/NYT.js:27]: Does this resolve the fixme or not yet?

File: package.json (hunk 1 above)
Comment (hyperknot) [reply package.json:44]: Yes, having the releases in the main repo doesn't allow me to compress the repo with BFG Repo-Cleaner https://rtyley.github.io/bfg-repo-cleaner/

File: src/shared/scrapers/CA/index.js (hunk 3 above)
Comment (hyperknot) [reply src/shared/scrapers/CA/index.js:128]: Oh, that's a good idea to check with assert!

File: src/shared/scrapers/DE/DE-BB/index.js (hunk 4 above)
Comment (hyperknot) [reply src/shared/scrapers/DE/DE-BB/index.js:12]: Yes, `iso2:` is always 5 characters exactly.

Comment (hyperknot) [reply src/shared/scrapers/DE/DE-BB/index.js:12]: I mean those codes will definietly not change. The "perfect" one would be to use country-levels.js / splitId, but this was easier as I could just to a search-replace on the codebase.

File: src/shared/scrapers/ES/index.js (hunk 5 above)
Comment (hyperknot) [reply src/shared/scrapers/ES/index.js:11]: What is deburring?

File: src/shared/scrapers/JHU-USA.js (hunk 6 above)
Comment (hyperknot) [reply src/shared/scrapers/JHU-USA.js:12]: I agree!

File: src/shared/scrapers/US/NYT.js (hunk 9 above)
Comment (hyperknot) [reply src/shared/scrapers/US/NYT.js:27]: I have no idea what is the fixme about, TZ is a big project of @ryanblock.

File: src/shared/scrapers/US/KY/index.js (hunk 8 above)
Comment (hyperknot) [reply src/shared/scrapers/US/KY/index.js:24]: In maintainers we could change all those back to USA I think. Right now the search-replace was safe and consistent and I preferred the safety of that.

File: src/shared/scrapers/ES/index.js (hunk 5 above)
Comment (ryanblock) [reply src/shared/scrapers/ES/index.js:11]: tl;dr nearest ascii conversion

File: src/shared/scrapers/US/NYT.js (hunk 9 above)
Comment (ryanblock) [reply src/shared/scrapers/US/NYT.js:27]: It does not, but that is in flight!

File: src/shared/scrapers/US/NYT.js [hunk 10]
```diff
@@ -25,26 +25,25 @@ const scraper = {
...
-    const fallback = process.env.USE_ISO_DATETIME ? new Date(datetime.now.at('America/New_York')) : datetime.getDate();
-    let scrapeDate = process.env.SCRAPE_DATE ? new Date(`${process.env.SCRAPE_DATE} 12:00:00`) : fallback;
+    let scrapeDate = process.env.SCRAPE_DATE
+      ? new Date(`${process.env.SCRAPE_DATE} 12:00:00`)
+      : new Date(datetime.now.at('America/New_York'));
```
Comment (ryanblock) [inline src/shared/scrapers/US/NYT.js:30]: Not sure I'd suggest using this yet, may be safer to stick with `datetime.getDate()` for the time being!

**Comment by lazd** [comment]
Merging this now, @camjc @ryanblock please file issues for the comments above. Thanks!

**Pull request merged**